_posts/2021-05-20-implementing-custom-facts-for-foreman-discovery.md

```diff
@@ -0,0 +1,109 @@
+
+```
+MyProjectDirectory
+ ├── autostart.d
+ ├── bin
+ ├── facts
+ │   └── nuc_position_facts.rb
+ └── lib
+     └── ruby
+```
+
+As you can see, to generate the facts we want, we implement a bit of ruby code that is consumed by facter. This ruby file is tucked under the facts directory. In the example, we simply created a case statement to map the MAC address to the position where the board is installed within the chassis. A simple row and column model like a spreadsheet is used in this example. All the systems connected by risers are a “Stack” and each is positioned from the connecting side. The ruby code looks like this:
+
+```ruby
+# nuc_position_facts.rb
+
+# Assign a position based on the MAC Address
+Facter.add("chassis_position") do
+  setcode do
+    mac = Facter.value('discovery_bootif')
+    case mac
+    when "ff:ee:dd:cc:bb:31"
+      "Stack3 NUC1"
+    when "ff:ee:dd:cc:bb:32"
+      "Stack3 NUC2"
+    when "ff:ee:dd:cc:bb:33"
+      "Stack3 NUC3"
+...
+...
+    else
+      mac
+    end
+  end
+
+```
+
+This isn’t the only thing that we can add to the Foreman Discovery Image. There are a number of other extensions that can be made use of. In the `bin` directory, binaries can be included that are able to be used in scripts to perform actions required at discovery time. In the `autostart.d` directory shell scripts can be added that use the included binaries. The `lib` and `lib/ruby` directories are added to the `LD_LIBRARY_PATH` and `RUBYLIB` path respectively so that any included binaries can be executed properly. Please see the [Foreman Provisioning documentation](https://docs.theforeman.org/nightly/Provisioning_Guide/index-foreman-el.html#Extending_the_Discovery_Image) for more details.
+
+With the example file built, it is time to deploy the code to the Foreman server to make use of it with discovery. In the `MyProjectDirectory`, zip the contents of the directory tree to a file with:
+
+```
+zip -r discovery_facts.zip .
+```
+
+Copy the resulting zip file to the foreman server and place it in `/var/lib/tftpboot/boot`
+
+```
+scp discovery_facts.zip root@myforemanhost.example.com:/var/lib/tftpboot/boot/
+```
+
+
+Now we need to modify a provisioning template to tell the discovery image that there is something additional that it needs to include when it starts. Navigate to the provisioning templates page on your foreman server and search for the `pxelinux_discovery` snippet. You can unlock the snippet to edit it or clone it. For simplicity we will unlock the templates for editing.
+
+Edit the `APPEND` line in `pxelinux_discover`y provisioning snippet on your foreman server to include a parameter that references the zip file. You can use more than one zip file to ease some of the management burden. The file can be located anywhere under `tftpboot`, the `fdi.zips` parameter provides the relative path. It can be beneficial to have separate directories for your code. The line for the example look similar to this:
```

Comment:
```suggestion
Edit the `APPEND` line in `pxelinux_discovery` provisioning snippet on your foreman server to include a parameter that references the zip file. You can use more than one zip file to ease some of the management burden. The file can be located anywhere under `tftpboot`, the `fdi.zips` parameter provides the relative path. It can be beneficial to have separate directories for your code. The line for the example look similar to this:
```